File: src/aggregator.py
"""
Financial aggregation and statistics computation.

This module computes totals, statistics, and summaries from categorized
transactions.
"""

from dataclasses import dataclass
from typing import Dict, List, Tuple
from src.parser import Transaction


@dataclass
class FinancialSummary:
    """Contains computed financial statistics."""

    total_income: float
    total_expenses: float
    net_change: float
    spending_by_category: Dict[str, float]
    largest_expense: Tuple[str, float] or None
    transaction_count: int
    uncategorized_transactions: List[Transaction]


class FinancialAggregator:
    """Computes financial statistics from categorized transactions."""

    def __init__(self, categorized_transactions: Dict[Transaction, str]):
        """
        Initialize aggregator with categorized transactions.

        Args:
            categorized_transactions: Dict mapping transactions to categories
        """
        self.categorized_transactions = categorized_transactions

    def aggregate(self) -> FinancialSummary:
        """
        Compute all financial statistics.

        Returns:
            FinancialSummary object with computed statistics
        """
        total_income = 0.0
        total_expenses = 0.0
        spending_by_category: Dict[str, float] = {}
        largest_expense: Tuple[str, float] = None
        uncategorized_transactions: List[Transaction] = []

        # Process each transaction
        for transaction, category in self.categorized_transactions.items():
            # Track income and expenses
            if transaction.is_income():
                total_income += transaction.amount
            else:
                total_expenses += abs(transaction.amount)

            # Track spending by category
            if category not in spending_by_category:
                spending_by_category[category] = 0.0
            spending_by_category[category] += abs(transaction.amount)

            # Track largest expense
            if transaction.is_expense():
                expense_amount = abs(transaction.amount)
                if largest_expense is None or expense_amount > largest_expense[1]:
                    largest_expense = (transaction.description, expense_amount)

            # Track uncategorized transactions
            if category == "Uncategorized":
                uncategorized_transactions.append(transaction)

        # Calculate net change
        net_change = total_income - total_expenses

        return FinancialSummary(
            total_income=total_income,
            total_expenses=total_expenses,
            net_change=net_change,
            spending_by_category=spending_by_category,
            largest_expense=largest_expense,
            transaction_count=len(self.categorized_transactions),
            uncategorized_transactions=uncategorized_transactions
        )

File: src/aggregator.js
export function aggregateData(categorizedTransactions) {
    let totalIncome = 0;
    let totalExpenses = 0;
    const categoryTotals = {};
    let largestExpense = null;

    console.log('Aggregating data for', categorizedTransactions.length, 'items');

    categorizedTransactions.forEach((item, index) => {
        if (!item) {
            console.warn(`Item ${index} is null/undefined`);
            return;
        }

        const { transaction, category } = item;

        if (!transaction) {
            console.warn(`Item ${index} has no transaction:`, item);
            return;
        }

        if (typeof transaction.amount === 'undefined') {
            console.warn(`Item ${index} transaction has no amount:`, transaction);
            return;
        }

        const amount = transaction.amount;

        if (amount > 0) {
            totalIncome += amount;
        } else {
            totalExpenses += Math.abs(amount);
        }

        if (!categoryTotals[category]) {
            categoryTotals[category] = 0;
        }
        categoryTotals[category] += Math.abs(amount);

        if (amount < 0) {
            const expenseAmount = Math.abs(amount);
            if (!largestExpense || expenseAmount > largestExpense.amount) {
                largestExpense = {
                    description: transaction.description,
                    amount: expenseAmount
                };
            }
        }
    });

    const netChange = totalIncome - totalExpenses;

    const sortedCategories = Object.entries(categoryTotals)
        .sort((a, b) => b[1] - a[1]);

    return {
        totalIncome,
        totalExpenses,
        netChange,
        categoryTotals: sortedCategories,
        largestExpense,
        transactionCount: categorizedTransactions.length
    };
}

export function groupTransactionsByMonth(categorizedTransactions) {
    const monthGroups = {};

    categorizedTransactions.forEach(item => {
        if (!item || !item.transaction || !item.transaction.date) {
            return;
        }

        const dateParts = item.transaction.date.split('/');
        if (dateParts.length < 3) {
            return;
        }

        const [month, day, year] = dateParts;
        const monthKey = `${year}-${month.padStart(2, '0')}`;

        if (!monthGroups[monthKey]) {
            monthGroups[monthKey] = [];
        }
        monthGroups[monthKey].push(item);
    });

    const months = Object.keys(monthGroups).sort();

    return months.map(monthKey => {
        const monthTransactions = monthGroups[monthKey];
        return {
            month: monthKey,
            transactions: monthTransactions.map(item => ({
                ...item.transaction,
                category: item.category
            })),
            summary: aggregateData(monthTransactions)
        };
    });
}
